Unity/Assets/NativeScript/Bindings.cs

```diff
@@ -24,44 +25,36 @@ public static class Bindings
 		// Holds objects and provides handles to them in the form of ints
 		public static class ObjectStore
 		{
+			// Lookup handles by object.
+			static Dictionary<object, int> objectHandleCache;
+
+			// Stack of available handles.
+			static Stack<int> handles;
+			
 			// Stored objects. The first is never used so 0 can be "null".
 			static object[] objects;
 			
-			// Stack of available handles
-			static int[] handles;
-			
-			// Hash table of stored objects to their handles.
-			static object[] keys;
-			static int[] values;
-			
-			// Index of the next available handle
-			static int nextHandleIndex;
-			
 			// The maximum number of objects to store. Must be positive.
 			static int maxObjects;
 			
 			public static void Init(int maxObjects)
 			{
 				ObjectStore.maxObjects = maxObjects;
+				objectHandleCache = new Dictionary<object, int> (maxObjects);	
+				handles = new Stack<int> (maxObjects);
 				
 				// Initialize the objects as all null plus room for the
 				// first to always be null.
 				objects = new object[maxObjects + 1];
 
 				// Initialize the handles stack as 1, 2, 3, ...
-				handles = new int[maxObjects];
 				for (
 					int i = 0, handle = maxObjects;
 					i < maxObjects;
 					++i, --handle)
 				{
-					handles[i] = handle;
+					handles.Push(handle);
```

Comment:
The switch to `Dictionary` is solving a performance issue, the purpose of this PR, but the use of `Stack` actually decreases performance due to the overhead included in functions like `Push` here. Instead of just assigning to an array element, it also checks if the stack is full. Since we know that's not possible, the check can be safely omitted. Yet the use of `Stack` means that check will be needlessly performed `maxObjects` times. `Stack` also doesn't remove much complexity from the code. So please revert the switch to `Stack`.

Reply:
Thanks for reviewing this!

This is a good point, I'll revert the handle stack back to using what you had previously.